[pybind11] feat(types) numpy.typing.NDArray

Based on https://github.com/pybind/pybind11/pull/5212 which had:
  Author: Michael Carlstrom <rmc@carlstrom.com>
  Co-authored-by: Tim Ohliger <ohliger.tim@gmail.com>

Due to rebasing challenges, we just keep a small part of that change,
namely we use the "return type" annotation spelling even for arguments,
which is a bit too strict (we also allow lists) but is better than a
wholly broken annotation.

--- include/pybind11/eigen/matrix.h
+++ include/pybind11/eigen/matrix.h
@@ -225,10 +225,11 @@ struct EigenProps {
         = !show_c_contiguous && show_order && requires_col_major;
 
     static constexpr auto descriptor
-        = const_name("numpy.ndarray[") + npy_format_descriptor<Scalar>::name + const_name("[")
+        = const_name("typing.Annotated[numpy.typing.NDArray[")
+          + npy_format_descriptor<Scalar>::name + const_name("], \"[")
           + const_name<fixed_rows>(const_name<(size_t) rows>(), const_name("m")) + const_name(", ")
-          + const_name<fixed_cols>(const_name<(size_t) cols>(), const_name("n")) + const_name("]")
+          + const_name<fixed_cols>(const_name<(size_t) cols>(), const_name("n")) + const_name("]\"")
           +
           // For a reference type (e.g. Ref<MatrixXd>) we have other constraints that might need to
           // be satisfied: writeable=True (for a mutable reference), and, depending on the map's
           // stride options, possibly f_contiguous or c_contiguous.  We include them in the
